File: WePeiYang/lostfound/src/main/java/com/twtstudio/tjwhm/lostfound/mylist/MylistPresenterImpl.java
package com.twtstudio.tjwhm.lostfound.mylist;

import com.twt.wepeiyang.commons.network.RetrofitProvider;
import com.twt.wepeiyang.commons.network.RxErrorHandler;
import com.twtstudio.tjwhm.lostfound.base.BaseBean;
import com.twtstudio.tjwhm.lostfound.base.CallbackBean;

import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

/**
 * Created by tjwhm on 2017/7/7.
 **/

public class MylistPresenterImpl implements MylistContract.MylistPresenter {
    MylistContract.MylistView mylistView;
    MylistApi mylistApi;

    public MylistPresenterImpl(MylistContract.MylistView mylistView) {
        this.mylistView = mylistView;
    }

    @Override
    public void setMylistData(MylistBean mylistBean) {
        mylistView.setMylistData(mylistBean);
    }

    @Override
    public void loadMylistData(String lostOrFound, int page) {
        mylistApi = RetrofitProvider.getRetrofit().create(MylistApi.class);
        mylistApi.loadMylistData(lostOrFound, String.valueOf(page))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setMylistData, new RxErrorHandler());
    }

    @Override
    public void turnStatus(int id) {
        mylistApi = RetrofitProvider.getRetrofit().create(MylistApi.class);
        mylistApi.turnStatus(String.valueOf(id))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::turnStatusSuccessCallBack, new RxErrorHandler());
    }

    @Override
    public void turnStatusSuccessCallBack(CallbackBean callbackBean) {
        mylistView.turnStatusSuccessCallBack();
    }
}

File: WePeiYang/lostfound2/src/main/java/com/yookiely/lostfond2/mylist/MyListPresenterImpl.kt
package com.yookiely.lostfond2.mylist

import android.util.Log
import android.widget.Toast
import com.yookiely.lostfond2.service.LostFoundService
import com.yookiely.lostfond2.service.MyListDataOrSearchBean
import com.twt.wepeiyang.commons.experimental.extensions.QuietCoroutineExceptionHandler
import com.twt.wepeiyang.commons.experimental.extensions.awaitAndHandle
import com.twt.wepeiyang.commons.experimental.network.CommonBody
import es.dmoral.toasty.Toasty
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class MyListPresenterImpl(private val myListView: MyListService.MyListView) : MyListService.MyListPresenter {
//    lateinit var mylistApi: LostFoundService

    override fun setMyListData(myListBean: List<MyListDataOrSearchBean>) {
        myListView.setMyListData(myListBean)
    }

    override fun loadMyListData(lostOrFound: String, page: Int) {
        GlobalScope.launch(Dispatchers.Main + QuietCoroutineExceptionHandler) {
            val myList: CommonBody<List<MyListDataOrSearchBean>>? = LostFoundService.getMyList(lostOrFound, page).awaitAndHandle {
                it.printStackTrace()
                Log.d("mylist",it.message)
            }
            if (myList == null) {
                val myListFragement = myListView as MyListFragment
                Toasty.error(myListFragement.context!!, "你网络崩啦，拿不到数据啦", Toast.LENGTH_LONG, true).show()
            } else {
                if (myList.error_code == -1) {
                    val dataBean: MutableList<MyListDataOrSearchBean> = mutableListOf()
                    for (temp in myList.data!!) {
                        if (lostOrFound == "found" && temp.type == 1) {
                            dataBean.add(temp)
                        } else {
                            if (temp.type == 0 && lostOrFound == "lost") {
                                dataBean.add(temp)
                            }
                        }
                    }
                    setMyListData(dataBean)
                }
            }
        }
    }

    override fun turnStatus(id: Int) {
        GlobalScope.launch(Dispatchers.Main + QuietCoroutineExceptionHandler) {
            val myList: CommonBody<String> = LostFoundService.turnStatus(id.toString()).await()

            if (myList.error_code == -1) {
                turnStatusSuccessCallBack(myList.data!!)
            }
        }
    }

    override fun turnStatusSuccessCallBack(callbackBean: String) {
        myListView.turnStatusSuccessCallBack()
    }
}